language-plutus-core/src/Language/PlutusCore/Mark.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Language.PlutusCore.Mark
+    ( markNonFresh
+    , markNonFreshBelow
+    , markNonFreshTerm
+    , markNonFreshType
+    , markNonFreshProgram
+    ) where
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Type
+
+import           Control.Lens
+import           Control.Monad.State
+import           Data.Functor.Foldable
+import           Data.Maybe                (fromMaybe)
+import qualified Data.Set                  as Set
+
+-- | Marks all the 'Unique's in a type as used, so they will not be generated in future. Useful if you
+-- have a type which was not generated in 'Quote'.
+markNonFreshType
+    :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
+    => Type tyname ann
+    -> m ()
+markNonFreshType = markNonFreshMax . collectTypeUniques
```

Comment:
Extracted various `mark*` functions into their own module to untie the `Name` module from `Type`, `Term` and other such data types.

Reply:
You mean the `Quote` module? That's where they were before, and it seemed like a reasonable place. That does depend on `Name`, and it seems reasonable that it should. It also seems okay for it to depend on `Term` etc.

Reply:
Yeah, sorry, the `Quote` module.

> That's where they were before, and it seemed like a reasonable place. That does depend on `Name`, and it seems reasonable that it should. It also seems okay for it to depend on `Term` etc.

So I think that one thing that is questionable is defining uniques-related stuff and names-related stuff together. This module does not need any names, it only needs uniques and their infrastructure (`TypeUnique`, `HasUnique`, etc). It's the same in many other places: specific Plutus Core names are rarely needed, most of the time we're generic over names and just interested in their uniques.

I don't think the `Quote` module should depend on our specific data types. `Quote` is a pretty general thing and we have it in Midnight as well, so it might be nice to extract `Quote` + stuff related to uniques into a separate library. We can even have

```haskell
newtype ScopedUnique (scope :: k) = ScopedUnique Unique 
```

and the core renamer module (and the core equality checking module) in that library.

Then we define this in Plutus Core:

```haskell
data Scope = TypeScope | TermScope

type TypeUnique = ScopedUnique TypeScope
type TermUnique = ScopedUnique TermScope
```

and enjoy using the library.

All without talking whether we should define a default `Name` data type, whether it should have annotations, whether we should have `ScopedName`, etc.

Note that even function like `lookupName`, `insertByName`, etc are generic over the choice of name.

So I think keeping `Quote` free from `Term` etc is a nice idea and we should consider moving uniques to their own module and leave concrete names in `Name`, because we mostly care about names having a unique (and maybe a `Pretty` instance) and not about an annotation, a textual component, etc.

Reply:
Hmm, okay.

Perhaps what I really think is that we should have something like:
``
markNonFresh :: MonadQuote m => a -> Traversal' a Unique -> m ()
```
in `Quote` and then we could define the appropriate traversals in `Plated`.

Reply:
Ok, you're right. I started defining appropriate traversals, but wasn't able to finish that. I'll do that this Friday.

Reply:
Happy for this to happen in a follow-up.

Reply:
I'd prefer to have it here and I haven't addressed the other comments yet. But if @reactormonk prefers to have this as early as possible to reduce the amount of conflicts, then let's merge.

Reply:
I think getting most of the reshuffling out of the way soon is good. I think the question of where instances etc. live is more important.

Reply:
Ok, I removed the "do not merge" label. I personally still prefer to wait a bit, I don't need much time to finish this off, but feel free to merge and I'll address the comments in a follow-up.

Reply:
Updated. See [this commit](https://github.com/input-output-hk/plutus/pull/1704/commits/7f2abaa5778fdd85138488fadf3760c57490874d).